File: app/core/token_utils.py
import json
import math
import re
from typing import Any, Dict, List

# 近似按 4 字节 = 1 token 估算，避免引入额外 tokenizer 依赖
_APPROX_BYTES_PER_TOKEN = 4
# 为每条消息预留少量结构性开销，避免低估
_MESSAGE_TOKEN_OVERHEAD = 4
# 多模态图片的粗略 token 估算值，用于避免 base64 误判触发压缩
_IMAGE_TOKEN_ESTIMATE = 256
_DATA_URL_PATTERN = re.compile(
    r"data:image/[a-zA-Z0-9+.-]+;base64,[A-Za-z0-9+/=\r\n]+"
)


def _estimate_content_tokens(content: Any) -> int:
    """估算消息内容的 token 开销，兼容多模态数组结构。"""
    if not content:
        return 0
    if isinstance(content, str):
        if content.startswith("data:image/"):
            return _IMAGE_TOKEN_ESTIMATE
        if "data:image/" in content:
            # 将内嵌 data URL 替换为短占位，避免 base64 触发误判
            image_count = len(_DATA_URL_PATTERN.findall(content))
            stripped = _DATA_URL_PATTERN.sub("[image]", content)
            return approx_token_count(stripped) + image_count * _IMAGE_TOKEN_ESTIMATE
        return approx_token_count(content)
    if isinstance(content, list):
        total = 0
        for item in content:
            total += _estimate_content_tokens(item)
        return total
    if isinstance(content, dict):
        part_type = str(content.get("type", "")).lower()
        if part_type == "text":
            return approx_token_count(str(content.get("text", "") or ""))
        if part_type == "image_url" or "image_url" in content:
            return _IMAGE_TOKEN_ESTIMATE
        if "text" in content:
            return approx_token_count(str(content.get("text", "") or ""))
    try:
        return approx_token_count(json.dumps(content, ensure_ascii=False))
    except TypeError:
        return approx_token_count(str(content))


def approx_token_count(text: str) -> int:
    """粗略估算文本 token 数量。"""
    if not text:
        return 0
    return math.ceil(len(text) / _APPROX_BYTES_PER_TOKEN)


def trim_text_to_tokens(text: str, max_tokens: int, suffix: str = "...(truncated)") -> str:
    """按粗略 token 上限裁剪文本，避免单条内容撑爆上下文预算。"""
    if not text:
        return ""
    if max_tokens <= 0:
        return suffix
    if approx_token_count(text) <= max_tokens:
        return text
    suffix_text = suffix or ""
    suffix_tokens = approx_token_count(suffix_text)
    if max_tokens <= suffix_tokens:
        max_chars = max(1, max_tokens * _APPROX_BYTES_PER_TOKEN)
        return suffix_text[:max_chars]
    max_chars = max_tokens * _APPROX_BYTES_PER_TOKEN - len(suffix_text)
    trimmed = text[: max(0, max_chars)]
    return trimmed + suffix_text


def estimate_message_tokens(message: Dict[str, Any]) -> int:
    """估算单条消息的 token 开销。"""
    content = message.get("content", "")
    content_tokens = _estimate_content_tokens(content)
    # 思考类模型会返回 reasoning_content，需计入上下文与占用统计
    reasoning = message.get("reasoning_content") or message.get("reasoning") or ""
    if not isinstance(reasoning, str):
        try:
            reasoning = json.dumps(reasoning, ensure_ascii=False)
        except TypeError:
            reasoning = str(reasoning)
    return content_tokens + approx_token_count(reasoning) + _MESSAGE_TOKEN_OVERHEAD


def estimate_messages_tokens(messages: List[Dict[str, Any]]) -> int:
    """估算消息列表的总 token 开销。"""
    return sum(estimate_message_tokens(message) for message in messages)


def trim_messages_to_budget(messages: List[Dict[str, Any]], max_tokens: int) -> List[Dict[str, Any]]:
    """按预算保留最近消息，避免上下文持续膨胀。"""
    if not messages:
        return []
    if max_tokens <= 0:
        return [messages[-1]]

    selected: List[Dict[str, Any]] = []
    remaining = max_tokens
    for message in reversed(messages):
        cost = estimate_message_tokens(message)
        if cost <= remaining:
            selected.append(message)
            remaining -= cost
            continue
        if not selected:
            # 最后一条消息必须保留，确保上下文连续
            selected.append(message)
        break

    return list(reversed(selected))

File: src/core/token_utils.rs
// Token 估算工具：用于近似计算上下文占用并进行裁剪。
use regex::Regex;
use serde_json::Value;
use std::sync::OnceLock;
use tracing::error;

const APPROX_BYTES_PER_TOKEN: f64 = 4.0;
const MESSAGE_TOKEN_OVERHEAD: i64 = 4;
const IMAGE_TOKEN_ESTIMATE: i64 = 256;

pub fn approx_token_count(text: &str) -> i64 {
    if text.is_empty() {
        return 0;
    }
    ((text.len() as f64) / APPROX_BYTES_PER_TOKEN).ceil() as i64
}

pub fn trim_text_to_tokens(text: &str, max_tokens: i64, suffix: &str) -> String {
    if text.is_empty() {
        return String::new();
    }
    if max_tokens <= 0 {
        return suffix.to_string();
    }
    if approx_token_count(text) <= max_tokens {
        return text.to_string();
    }
    let suffix_text = suffix;
    let suffix_tokens = approx_token_count(suffix_text);
    if max_tokens <= suffix_tokens {
        let max_chars = (max_tokens.max(1) as f64 * APPROX_BYTES_PER_TOKEN) as usize;
        return suffix_text.chars().take(max_chars).collect();
    }
    let max_chars = (max_tokens as f64 * APPROX_BYTES_PER_TOKEN) as usize;
    let trimmed: String = text
        .chars()
        .take(max_chars.saturating_sub(suffix_text.len()))
        .collect();
    format!("{trimmed}{suffix_text}")
}

pub fn estimate_message_tokens(message: &Value) -> i64 {
    if !message.is_object() {
        return 0;
    }
    let content_tokens = estimate_content_tokens(message.get("content").unwrap_or(&Value::Null));
    let reasoning = message
        .get("reasoning_content")
        .or_else(|| message.get("reasoning"))
        .unwrap_or(&Value::Null);
    let reasoning_tokens = match reasoning {
        Value::String(text) => approx_token_count(text),
        Value::Array(_) | Value::Object(_) => approx_token_count(&reasoning.to_string()),
        _ => 0,
    };
    content_tokens + reasoning_tokens + MESSAGE_TOKEN_OVERHEAD
}

pub fn estimate_messages_tokens(messages: &[Value]) -> i64 {
    messages.iter().map(estimate_message_tokens).sum()
}

pub fn trim_messages_to_budget(messages: &[Value], max_tokens: i64) -> Vec<Value> {
    if messages.is_empty() {
        return Vec::new();
    }
    if max_tokens <= 0 {
        return vec![messages[messages.len() - 1].clone()];
    }
    let mut selected: Vec<Value> = Vec::new();
    let mut remaining = max_tokens;
    for message in messages.iter().rev() {
        let cost = estimate_message_tokens(message);
        if cost <= remaining {
            selected.push(message.clone());
            remaining -= cost;
            continue;
        }
        if selected.is_empty() {
            selected.push(message.clone());
        }
        break;
    }
    selected.reverse();
    selected
}

pub fn estimate_content_tokens(content: &Value) -> i64 {
    match content {
        Value::Null => 0,
        Value::String(text) => estimate_string_tokens(text),
        Value::Array(items) => items.iter().map(estimate_content_tokens).sum(),
        Value::Object(map) => {
            let part_type = map
                .get("type")
                .and_then(Value::as_str)
                .unwrap_or("")
                .to_lowercase();
            if part_type == "text" {
                return approx_token_count(map.get("text").and_then(Value::as_str).unwrap_or(""));
            }
            if part_type == "image_url" || map.contains_key("image_url") {
                return IMAGE_TOKEN_ESTIMATE;
            }
            if let Some(text) = map.get("text").and_then(Value::as_str) {
                return approx_token_count(text);
            }
            approx_token_count(&content.to_string())
        }
        _ => approx_token_count(&content.to_string()),
    }
}

fn estimate_string_tokens(text: &str) -> i64 {
    if text.starts_with("data:image/") {
        return IMAGE_TOKEN_ESTIMATE;
    }
    if text.contains("data:image/") {
        let Some(re) = data_url_regex() else {
            return approx_token_count(text);
        };
        let matches = re.find_iter(text).count() as i64;
        let stripped = re.replace_all(text, "[image]");
        return approx_token_count(&stripped) + matches * IMAGE_TOKEN_ESTIMATE;
    }
    approx_token_count(text)
}

fn data_url_regex() -> Option<&'static Regex> {
    static REGEX: OnceLock<Option<Regex>> = OnceLock::new();
    REGEX
        .get_or_init(|| {
            match Regex::new(r"data:image/[a-zA-Z0-9+.-]+;base64,[A-Za-z0-9+/=\r\n]+") {
                Ok(regex) => Some(regex),
                Err(err) => {
                    error!("invalid token_utils data url regex: {err}");
                    None
                }
            }
        })
        .as_ref()
}
